File: lib/services/ai_memory_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'local_storage.dart';
import 'couple_service.dart';

class AiMemory {
  String key;
  String value;
  String category;
  DateTime updatedAt;

  AiMemory({
    required this.key,
    required this.value,
    required this.category,
    DateTime? updatedAt,
  }) : updatedAt = updatedAt ?? DateTime.now();

  Map<String, dynamic> toJson() => {
    'key': key,
    'value': value,
    'category': category,
    'updatedAt': updatedAt.toIso8601String(),
  };

  factory AiMemory.fromJson(Map<String, dynamic> json) => AiMemory(
    key: json['key'] as String,
    value: json['value'] as String,
    category: json['category'] as String,
    updatedAt: DateTime.tryParse(json['updatedAt'] as String? ?? '') ?? DateTime.now(),
  );
}

class AiMemoryService extends ChangeNotifier {
  static final AiMemoryService _instance = AiMemoryService._();
  factory AiMemoryService() => _instance;
  AiMemoryService._();

  final List<AiMemory> _memories = [];
  bool _loaded = false;

  List<AiMemory> get memories => List.unmodifiable(_memories);
  bool get isLoaded => _loaded;

  // Memoria predefinida de la pareja
  static const _defaults = {
    'nombres': 'ambos',
    'reaccion': 'carinosa',
  };

  Future<void> load() async {
    if (_loaded) return;
    final cached = LocalStorage().getString('ai_memories');
    if (cached != null) {
      try {
        final list = jsonDecode(cached) as List;
        _memories.addAll(list.map((e) => AiMemory.fromJson(e as Map<String, dynamic>)));
      } catch (_) {}
    }
    // Cargar desde Firestore
    try {
      final doc = await CoupleService().ref.collection('ai').doc('memoria').get();
      if (doc.exists) {
        final data = doc.data() as Map<String, dynamic>;
        for (final entry in data.entries) {
          if (entry.value is Map) {
            final m = AiMemory(
              key: entry.key,
              value: entry.value['value'] as String? ?? '',
              category: entry.value['category'] as String? ?? 'general',
            );
            _memories.add(m);
          }
        }
      }
    } catch (_) {}
    _loaded = true;
    notifyListeners();
  }

  Future<void> setMemory(String key, String value, {String category = 'general'}) async {
    _memories.removeWhere((m) => m.key == key);
    _memories.add(AiMemory(key: key, value: value, category: category));
    _saveToLocal();
    // Sync to Firestore
    try {
      await CoupleService().ref.collection('ai').doc('memoria').set({
        key: {'value': value, 'category': category, 'updatedAt': DateTime.now().toIso8601String()},
      }, SetOptions(merge: true));
    } catch (_) {}
    notifyListeners();
  }

  String? getMemory(String key) {
    try {
      return _memories.firstWhere((m) => m.key == key).value;
    } catch (_) {
      return null;
    }
  }

  String buildContextPrompt() {
    final buf = StringBuffer();
    buf.writeln('Informacion de la pareja:');
    for (final m in _memories) {
      buf.writeln('- ${m.key}: ${m.value}');
    }
    return buf.toString();
  }

  void _saveToLocal() {
    final json = jsonEncode(_memories.map((m) => m.toJson()).toList());
    LocalStorage().setString('ai_memories', json);
  }
}

File: iphone app/Novios/Services/AIMemoryService.swift
import Foundation
import Combine
import FirebaseFirestore

struct AIMemory: Codable {
    var key: String
    var value: String
    var category: String
    var updatedAt: String
}

class AIMemoryService: ObservableObject {
    static let shared = AIMemoryService()

    @Published var memories: [AIMemory] = []
    private let db = Firestore.firestore()
    private let defaults = UserDefaults.standard

    private var coupleId: String { CoupleService.coupleId }

    func load() {
        if let data = defaults.data(forKey: "ai_memories"),
           let cached = try? JSONDecoder().decode([AIMemory].self, from: data) {
            memories = cached
        }
        // Load from Firestore
        db.collection("parejas").document(coupleId).collection("ai").document("memoria").getDocument { [weak self] snap, _ in
            guard let data = snap?.data() else { return }
            for (key, val) in data {
                if let dict = val as? [String: Any],
                   let value = dict["value"] as? String,
                   let category = dict["category"] as? String {
                    let mem = AIMemory(key: key, value: value, category: category, updatedAt: dict["updatedAt"] as? String ?? "")
                    self?.memories.append(mem)
                }
            }
            self?.saveLocal()
        }
    }

    func setMemory(key: String, value: String, category: String = "general") {
        memories.removeAll { $0.key == key }
        let mem = AIMemory(key: key, value: value, category: category, updatedAt: ISO8601DateFormatter().string(from: Date()))
        memories.append(mem)
        saveLocal()
        // Sync to Firestore
        db.collection("parejas").document(coupleId).collection("ai").document("memoria").setData([
            key: ["value": value, "category": category, "updatedAt": mem.updatedAt]
        ], merge: true)
    }

    func getMemory(_ key: String) -> String? {
        memories.first { $0.key == key }?.value
    }

    func buildContextPrompt() -> String {
        memories.map { "- \($0.key): \($0.value)" }.joined(separator: "\n")
    }

    private func saveLocal() {
        if let data = try? JSONEncoder().encode(memories) {
            defaults.set(data, forKey: "ai_memories")
        }
    }
}
